.github/workflows/published-charms-tests.yaml

```diff
@@ -0,0 +1,124 @@
+          - charm-repo: canonical/route53-lego-k8s-operator
+          - charm-repo: canonical/wordpress-k8s-operator
+          - charm-repo: canonical/temporal-k8s-operator
+          - charm-repo: canonical/jenkins-agent-operator
+          - charm-repo: canonical/seldon-core-operator
+          - charm-repo: canonical/nginx-ingress-integrator-operator
+          - charm-repo: canonical/oathkeeper-operator
+          - charm-repo: canonical/jenkins-agent-k8s-operator
+          - charm-repo: canonical/grafana-agent-k8s-operator
+          - charm-repo: canonical/namecheap-lego-k8s-operator
+          - charm-repo: canonical/superset-k8s-operator
+          - charm-repo: canonical/identity-platform-login-ui-operator
+          - charm-repo: canonical/temporal-admin-k8s-operator
+          - charm-repo: canonical/dex-auth-operator
+          - charm-repo: canonical/temporal-ui-k8s-operator
+          - charm-repo: canonical/temporal-worker-k8s-operator
+          - charm-repo: canonical/content-cache-k8s-operator
+          - charm-repo: canonical/ranger-k8s-operator
+          - charm-repo: canonical/oauth2-proxy-k8s-operator
+          - charm-repo: canonical/livepatch-k8s-operator
+          - charm-repo: canonical/self-signed-certificates-operator
+          - charm-repo: canonical/saml-integrator-operator
+          - charm-repo: canonical/smtp-integrator-operator
+          - charm-repo: canonical/manual-tls-certificates-operator
+          - charm-repo: canonical/hardware-observer-operator
+    steps:
+      - name: Checkout the ${{ matrix.charm-repo }} repository
+        uses: actions/checkout@v4
+        with:
+          repository: ${{ matrix.charm-repo }}
+
+      - name: Checkout the operator repository
+        uses: actions/checkout@v4
+        with:
+          path: myops
+
+      - name: Install patch dependencies
+        run: pip install poetry~=1.6
+
+      - name: Update 'ops' dependency in test charm to latest
+        run: |
+          rm -rf myops/test
+          if [ -e "test-requirements.txt" ]; then
+            sed -i -e "/^ops[ ><=]/d" -e "/canonical\/operator/d" -e "/#egg=ops/d" test-requirements.txt
+            echo -e "\ngit+$GITHUB_SERVER_URL/$GITHUB_REPOSITORY@$GITHUB_SHA#egg=ops" >> test-requirements.txt
+          fi
+          if [ -e "requirements-charmcraft.txt" ]; then
+            sed -i -e "/^ops[ ><=]/d" -e "/canonical\/operator/d" -e "/#egg=ops/d" requirements-charmcraft.txt
+            echo -e "\ngit+$GITHUB_SERVER_URL/$GITHUB_REPOSITORY@$GITHUB_SHA#egg=ops" >> requirements-charmcraft.txt
+          fi
+          if [ -e "requirements.txt" ]; then
+            sed -i -e "/^ops[ ><=]/d" -e "/canonical\/operator/d" -e "/#egg=ops/d" requirements.txt
+            echo -e "\ngit+$GITHUB_SERVER_URL/$GITHUB_REPOSITORY@$GITHUB_SHA#egg=ops" >> requirements.txt
+          else
+            sed -i -e "s/^ops[ ><=].*/ops = {path = \"myops\"}/" pyproject.toml
```

Comment:
> How do we know `pyproject.toml` exists in this branch? If there's no requirements.txt we assume pyproject -- is that safe enough?

If the charm is using poetry or uv then `pyproject.toml` should exist and have the dependencies. I'm not entirely sure for other tools like hatch.

I guess this should really look for `poetry.lock` and do the pyproject.toml edit plus poetry lock if it exists, and fail otherwise, and then if there are failures we can add support for other tools if published charms adopt them.

For what it's worth, the dependency working group's spec currently requires a requirements.txt.